# Importing the libraries

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns 

# Loading the dataset

In [2]:
data = pd.read_csv('learningSet.csv')
data.shape

(95412, 481)

In [3]:
data.head()

ODATEDW OSOURCE  TCODE STATE    ZIP MAILCODE PVASTATE   DOB NOEXCH  \
0     8901     GRI      0    IL  61081                    3712      0   
1     9401     BOA      1    CA  91326                    5202      0   
2     9001     AMH      1    NC  27017                       0      0   
3     8701     BRY      0    CA  95953                    2801      0   
4     8601              0    FL  33176                    2001      0   

  RECINHSE RECP3 RECPGVG RECSWEEP MDMAUD DOMAIN CLUSTER   AGE AGEFLAG  \
0                                   XXXX     T2      36  60.0           
1                                   XXXX     S1      14  46.0       E   
2                                   XXXX     R2      43   NaN           
3                                   XXXX     R2      44  70.0       E   
4        X     X                    XXXX     S2      16  78.0       E   

  HOMEOWNR CHILD03 CHILD07 CHILD12 CHILD18  NUMCHLD  INCOME GENDER  WEALTH1  \
0                                               NaN     NaN      F      NaN   
1        H                               M      1.0     6.0      M      9.0   
2        U                                      NaN     3.0      M      1.0   
3        U                                      NaN     1.0      F      4.0   
4        H                                      1.0     3.0      F      2.0   

   HIT  MBCRAFT  MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  MAGMALE  \
0    0      NaN       NaN      NaN       NaN      NaN     NaN      NaN   
1   16      0.0       0.0      3.0       1.0      1.0     1.0      0.0   
2    2      0.0       0.0      1.0       0.0      0.0     0.0      0.0   
3    2      0.0       0.0      0.0       0.0      0.0     0.0      0.0   
4   60      1.0       0.0      9.0       0.0      4.0     1.0      0.0   

   PUBGARDN  PUBCULIN  PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  PUBOPP DATASRCE  \
0       NaN       NaN      NaN       NaN       NaN       NaN     NaN            
1       0.0       0.0      2.0       0.0       3.0       0.0     0.0        3   
2       0.0       0.0      1.0       0.0       0.0       0.0     0.0        3   
3       0.0       0.0      0.0       0.0       0.0       0.0     1.0        3   
4       0.0       0.0      4.0       0.0       1.0       0.0     1.0        3   

   MALEMILI  MALEVET  VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV SOLP3  \
0         0       39        34        18        10         2       1         
1         0       15        55        11         6         2       1         
2         0       20        29        33         6         8       1         
3         0       23        14        31         3         0       3         
4         1       28         9        53        26         3       2         

  SOLIH MAJOR  WEALTH2 GEOCODE COLLECT1 VETERANS BIBLE CATLG HOMEE PETS  \
0                  5.0                                                    
1                  9.0       2                                            
2                  1.0                                                    
3                  0.0                                                    
4    12            NaN                               Y     Y              

  CDPLAY STEREO PCOWNERS PHOTO CRAFTS FISHER GARDENIN BOATS WALKER KIDSTUFF  \
0                                                                             
1                                                                             
2                                                                             
3                                                                             
4             Y                     Y               Y            Y            

  CARDS PLATES LIFESRC PEPSTRFL  POP901  POP902  POP903  POP90C1  POP90C2  \
0                             X     992     264     332        0       35   
1                                  3611     940     998       99        0   
2                             X    7001    2040    2669        0        2   
3              

A very large dataset with null values in columns and spaces in the values of some columns

# EDA

**Check for null values in all the columns**

In [4]:
nulls = data.isna().sum()
nulls.sort_values(ascending = False).head(80)

RDATE_5     95403
RAMNT_5     95403
RDATE_3     95170
RAMNT_3     95170
RDATE_4     95131
            ...  
ADATE_16    20364
ADATE_14    18867
ADATE_9     11245
ADATE_11    10422
NEXTDATE     9973
Length: 80, dtype: int64

It can be seen that some columns have a very high number of null values. Let's calculate the percentage of the null rows compared to the whole number of rows.

In [5]:
#Checking the percentage of null rows:
nulls_percentage = nulls * 100 / len(data)
nulls_percentage.sort_values(ascending = False)

RDATE_5    99.990567
RAMNT_5    99.990567
RDATE_3    99.746363
RAMNT_3    99.746363
RDATE_4    99.705488
             ...    
ETHC3       0.000000
ETHC2       0.000000
ETHC1       0.000000
HHD12       0.000000
TPE11       0.000000
Length: 481, dtype: float64

In [6]:
nulls_percentage_df = pd.DataFrame(nulls_percentage, columns = ['percentage'])
nulls_percentage_df

percentage
ODATEDW     0.000000
OSOURCE     0.000000
TCODE       0.000000
STATE       0.000000
ZIP         0.000000
...              ...
MDMAUD_R    0.000000
MDMAUD_F    0.000000
MDMAUD_A    0.000000
CLUSTER2    0.138347
GEOCODE2    0.138347

[481 rows x 1 columns]

**Exclude the following variables by looking at the definitions. Create a new empty list called drop_list. We will append this list and then drop all the columns in this list later:**

**OSOURCE - symbol definitions not provided, too many categories**

**ZIP CODE - we are including state already**

OSOURCE:

Origin Source
- Only 1rst 3 bytes are used
- Defaulted to 00000 for conversion
- Code indicating which mailing list the donor was originally acquired from
- A nominal or symbolic field.

In [7]:
data['OSOURCE'].nunique()

896

osource is the abbreviation for origin source . but it's not clear what it means and as can be seen, it has 896 categories.

In [8]:
data['ZIP'].nunique()

19938

ZIP column also has a large number of categories and its information has already been included in state column. So, it is redundant.

In [9]:
drop_list = data[['OSOURCE', 'ZIP']]

**Identify columns that over 85% missing values**



In [10]:
high_nulls = list(nulls_percentage_df[nulls_percentage_df['percentage'] > 0.85].index)
high_nulls

['AGE',
 'NUMCHLD',
 'INCOME',
 'WEALTH1',
 'MBCRAFT',
 'MBGARDEN',
 'MBBOOKS',
 'MBCOLECT',
 'MAGFAML',
 'MAGFEM',
 'MAGMALE',
 'PUBGARDN',
 'PUBCULIN',
 'PUBHLTH',
 'PUBDOITY',
 'PUBNEWFN',
 'PUBPHOTO',
 'PUBOPP',
 'WEALTH2',
 'ADATE_3',
 'ADATE_4',
 'ADATE_5',
 'ADATE_6',
 'ADATE_7',
 'ADATE_8',
 'ADATE_9',
 'ADATE_10',
 'ADATE_11',
 'ADATE_12',
 'ADATE_13',
 'ADATE_14',
 'ADATE_15',
 'ADATE_16',
 'ADATE_17',
 'ADATE_18',
 'ADATE_19',
 'ADATE_20',
 'ADATE_21',
 'ADATE_22',
 'ADATE_23',
 'ADATE_24',
 'RDATE_3',
 'RDATE_4',
 'RDATE_5',
 'RDATE_6',
 'RDATE_7',
 'RDATE_8',
 'RDATE_9',
 'RDATE_10',
 'RDATE_11',
 'RDATE_12',
 'RDATE_13',
 'RDATE_14',
 'RDATE_15',
 'RDATE_16',
 'RDATE_17',
 'RDATE_18',
 'RDATE_19',
 'RDATE_20',
 'RDATE_21',
 'RDATE_22',
 'RDATE_23',
 'RDATE_24',
 'RAMNT_3',
 'RAMNT_4',
 'RAMNT_5',
 'RAMNT_6',
 'RAMNT_7',
 'RAMNT_8',
 'RAMNT_9',
 'RAMNT_10',
 'RAMNT_11',
 'RAMNT_12',
 'RAMNT_13',
 'RAMNT_14',
 'RAMNT_15',
 'RAMNT_16',
 'RAMNT_17',
 'RAMNT_18',
 'RAMNT_19',


**Remove those columns from the dataframe**

In [11]:
#putting the columns that we want to keep in a list
data2 = []
for col in data.columns:
    if col not in high_nulls:
        data2.append(col)
data2

['ODATEDW',
 'OSOURCE',
 'TCODE',
 'STATE',
 'ZIP',
 'MAILCODE',
 'PVASTATE',
 'DOB',
 'NOEXCH',
 'RECINHSE',
 'RECP3',
 'RECPGVG',
 'RECSWEEP',
 'MDMAUD',
 'DOMAIN',
 'CLUSTER',
 'AGEFLAG',
 'HOMEOWNR',
 'CHILD03',
 'CHILD07',
 'CHILD12',
 'CHILD18',
 'GENDER',
 'HIT',
 'DATASRCE',
 'MALEMILI',
 'MALEVET',
 'VIETVETS',
 'WWIIVETS',
 'LOCALGOV',
 'STATEGOV',
 'FEDGOV',
 'SOLP3',
 'SOLIH',
 'MAJOR',
 'GEOCODE',
 'COLLECT1',
 'VETERANS',
 'BIBLE',
 'CATLG',
 'HOMEE',
 'PETS',
 'CDPLAY',
 'STEREO',
 'PCOWNERS',
 'PHOTO',
 'CRAFTS',
 'FISHER',
 'GARDENIN',
 'BOATS',
 'WALKER',
 'KIDSTUFF',
 'CARDS',
 'PLATES',
 'LIFESRC',
 'PEPSTRFL',
 'POP901',
 'POP902',
 'POP903',
 'POP90C1',
 'POP90C2',
 'POP90C3',
 'POP90C4',
 'POP90C5',
 'ETH1',
 'ETH2',
 'ETH3',
 'ETH4',
 'ETH5',
 'ETH6',
 'ETH7',
 'ETH8',
 'ETH9',
 'ETH10',
 'ETH11',
 'ETH12',
 'ETH13',
 'ETH14',
 'ETH15',
 'ETH16',
 'AGE901',
 'AGE902',
 'AGE903',
 'AGE904',
 'AGE905',
 'AGE906',
 'AGE907',
 'CHIL1',
 'CHIL2',
 'CHIL3',
 'AGEC1',


In [12]:
#removing the columns with high null values
data2_df = data.drop(columns = high_nulls)

In [13]:
data2_df

ODATEDW OSOURCE  TCODE STATE    ZIP MAILCODE PVASTATE   DOB NOEXCH  \
0         8901     GRI      0    IL  61081                    3712      0   
1         9401     BOA      1    CA  91326                    5202      0   
2         9001     AMH      1    NC  27017                       0      0   
3         8701     BRY      0    CA  95953                    2801      0   
4         8601              0    FL  33176                    2001      0   
...        ...     ...    ...   ...    ...      ...      ...   ...    ...   
95407     9601     ASE      1    AK  99504                       0      0   
95408     9601     DCD      1    TX  77379                    5001      0   
95409     9501     MBC      1    MI  48910                    3801      0   
95410     8601     PRV      0    CA  91320                    4005      0   
95411     8801     MCC      2    NC  28409                    1801      0   

      RECINHSE RECP3 RECPGVG RECSWEEP MDMAUD DOMAIN CLUSTER AGEFLAG HOMEOWNR  \
0                                       XXXX     T2      36                    
1                                       XXXX     S1      14       E        H   
2                                       XXXX     R2      43                U   
3                                       XXXX     R2      44       E        U   
4            X     X                    XXXX     S2      16       E        H   
...        ...   ...     ...      ...    ...    ...     ...     ...      ...   
95407                                   XXXX     C2      27                    
95408                                   XXXX     C1      24       E        H   
95409              X                    XXXX     C3      30                    
95410        X                          XXXX     C1      24       E        H   
95411        X             X            C1CM     C1      24       E        U   

      CHILD03 CHILD07 CHILD12 CHILD18 GENDER  HIT DATASRCE  MALEMILI  MALEVET  \
0                                          F    0                  0       39   
1                                   M      M   16        3         0       15   
2                                          M    2        3         0       20   
3                                          F    2        3         0       23   
4                                          F   60        3         1       28   
...       ...     ...     ...     ...    ...  ...      ...       ...      ...   
95407                                      M    0                 14       36   
95408                               M      M    1        3         0       31   
95409                                      M    0                  0       18   
95410                                      F    0        2         0       28   
95411                                      F    3        3         1       33   

       VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV SOLP3 SOLIH MAJOR  \
0            34        18        10         2       1                     
1            55        11         6         2       1                     
2            29        33         6         8       1                     
3            14        31         3         0       3                     
4             9        53        26         3       2          12         
...         ...       ...       ...       ...     ...   ...   ...   ...   
95407        47        11         7         8      13                     
95408        43        19         4         1       0                     
95409        46        20         7        23       0                     
95410        35        20         9         1       1          12         
95411        22        27         8        10       3          12     X   

      GEOCODE COLLECT1 VETERANS BIBLE CATLG HOMEE PETS CDPLAY STEREO PCOWNERS  \
0                                                                               
1           2                                                                   
2   

**Reduce the number of categories in the column GENDER. The column should only have either "M" for males, "F" for females, and "other" for all the rest**

In [14]:
data2_df['GENDER'].value_counts()

F    51277
M    39094
      2957
U     1715
J      365
C        2
A        2
Name: GENDER, dtype: int64

In [15]:
#We will first replace those null values using the code below:
data2_df['GENDER'] = data2_df['GENDER'].replace(' ', 'F')

In [16]:
#putting the value counts and categories in a dataframe
data_gender1 = pd.DataFrame(data2_df['GENDER'].value_counts().reset_index().values, columns=["GENDER", "number"])
data_gender1

GENDER number
0      F  54234
1      M  39094
2      U   1715
3      J    365
4      C      2
5      A      2

In [18]:
#defining a new dataframe for the categories with the frequency less thatn 1716
data_gender_new = data_gender1[data_gender1['number'] < 1716]

In [19]:

data_gender_new_1= list(data_gender_new['GENDER'])

In [24]:
data_gender_new_1

['U', 'J', 'C', 'A']

In [20]:
#defining a function for bucketing
def reduce_categories(x):
    if x in data_gender_new_1:
        return 'other'
    else:
        return x

In [21]:
#calling the function
data2_df['GENDER'] = data2_df['GENDER'].apply(reduce_categories)

In [22]:
data2_df['GENDER']

0        F
1        M
2        M
3        F
4        F
        ..
95407    M
95408    M
95409    M
95410    F
95411    F
Name: GENDER, Length: 95412, dtype: object

In [23]:
data2_df['GENDER'].value_counts()

F        54234
M        39094
other     2084
Name: GENDER, dtype: int64